src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1525,148 +1525,312 @@ static LogicalResult getPartiallyFlattenedSimdCode(
 // Utilities for Op fusion at lowering
 //===----------------------------------------------------------------------===//
 
-// Function pointer type for emitScalarOpFor<T>
+// Function pointer type for the emitScalarOpFor<T> of elementwise Ops.
 typedef mlir::Value (*EmitScalarFunc)(mlir::ConversionPatternRewriter &rewriter,
     mlir::Location loc, mlir::Operation *op, mlir::Type elementType,
     mlir::ArrayRef<mlir::Value> scalarOperands);
 
-// Variadic template to iterate all the fusible Ops
+// Utility class for Op fusion.
+// Start from the root op, which is being lowered as an Elementwise Op.
+// Following the def-use chain from the root op, a line of fusible ops are
+// idenitified. The fusible Ops have to be elementwise Op, and satisfy shape
+// and depedence requirement.
```

Comment:
Could it be more concrete here? It appears that not all elementwise ops are supported, e.g, two elementwise binary ops, like `Add -> Sub`. I would expect to see a list like:
```
- Pattern 1: unary op -> unary op
- Pattern 2: binary op -> unary op, .... requirements
- Pattern 3: unary op -> binary op, .... requirements
- other patterns ....
```

Reply:
Add and Sub are in variadic Op list. I just followed the ONNXElementwiseUnaryOpLowering<T>, ONNXElementwiseBinaryOpLowering<T>, and ONNXElementwiseVariadicOpLowering<T>.
They are all supported, except ClipOp. ClipOp is in ONNXElementwiseUnaryOpLowering, but it has three input and does not support broadcast. In fact, such op could be fused too.

Reply:
The condition for unary Op and binary Op are the same. But the unary is simpler  because it has only the input to be  fused, no extra loads for other input.